votingworks/vxsuite
# perf(logging): reduce peak memory usage in CDF export

## Overview
On a synthetic log file with 100K lines (34 MB) this change brought the total memory usage down 50% from 460 MB to 227 MB (plus 85 MB baseline of just libraries) and, as a side effect, brought the time down 37% from 25.7s to 16.1s.

## Demo Video or Screenshot

Times and memory values below were collected from a warmed-up test run on my M1 Max MBP.

### In-Memory String (`readFileSync`)

<details>
<summary>Full log</summary>
```
Exporting CDF log
Memory usage: heapUsed=571499792 external=651576
Memory usage: heapUsed=362438248 external=13233759
Memory usage: heapUsed=294056792 external=28438111
Memory usage: heapUsed=218824592 external=24243807
Memory usage: heapUsed=127427000 external=39448159
Memory usage: heapUsed=426224624 external=35909215
Memory usage: heapUsed=338728456 external=30928479
Memory usage: heapUsed=237197400 external=26209887
Memory usage: heapUsed=138215704 external=14380639
Memory usage: heapUsed=416056320 external=29322847
Memory usage: heapUsed=326098168 external=30928479
Memory usage: heapUsed=238230368 external=26078815
Memory usage: heapUsed=149328656 external=21229151
Memory usage: heapUsed=435210296 external=36695647
Memory usage: heapUsed=347125744 external=31845983
Memory usage: heapUsed=257678040 external=27127391
Memory usage: heapUsed=168878776 external=22408799
Memory usage: heapUsed=456208712 external=37482079
Memory usage: heapUsed=366227792 external=13102687
Memory usage: heapUsed=278333312 external=28044895
Memory usage: heapUsed=183075608 external=15191647
cdf-export: 25.683s
```
</details>

- Total time: **25.683s**
- Peak heap memory: **545 MB**
- Average heap memory: 287.8 MB
- Peak external memory: 37.6 MB
- Average external memory: 24.2 MB

### Streaming (`createReadStream`)

<details>
<summary>Full log</summary>
```
Exporting CDF log
Memory usage: heapUsed=189375728 external=6217879
Memory usage: heapUsed=289284920 external=11854015
Memory usage: heapUsed=124776032 external=2879327
Memory usage: heapUsed=138189320 external=3567455
Memory usage: heapUsed=100544776 external=3551071
Memory usage: heapUsed=208067160 external=9539423
Memory usage: heapUsed=136846008 external=4263775
Memory usage: heapUsed=230205064 external=8834911
Memory usage: heapUsed=327684880 external=14544735
Memory usage: heapUsed=183266512 external=5295967
Memory usage: heapUsed=280705600 external=11169631
Memory usage: heapUsed=103906088 external=1077087
Memory usage: heapUsed=201964704 external=6950751
Memory usage: heapUsed=302087528 external=12578655
Memory usage: heapUsed=142041424 external=3133279
Memory usage: heapUsed=232877488 external=8859487
cdf-export: 16.114s
```
</details>

- Total time: **16.114s**
- Peak heap memory: **312.5 MB**
- Average heap memory: 190.3 MB
- Peak external memory: 13.9 MB
- Average external memory: 6.8 MB

### Baseline (no export, load libraries only)

<details>
<summary>Full log</summary>
```
Memory usage: heapUsed=89442088 external=33042894
Memory usage: heapUsed=89446096 external=33042934
Memory usage: heapUsed=89448472 external=33042934
Memory usage: heapUsed=89450976 external=33042934
Memory usage: heapUsed=89453296 external=33042934
Memory usage: heapUsed=89456160 external=33042934
Memory usage: heapUsed=89458576 external=33042934
Memory usage: heapUsed=89461080 external=33042934
Memory usage: heapUsed=83540432 external=519775
Memory usage: heapUsed=83553464 external=519775
Memory usage: heapUsed=83555760 external=519775
Memory usage: heapUsed=83569800 external=519775
Memory usage: heapUsed=83779056 external=519775
Memory usage: heapUsed=83781728 external=519775
Memory usage: heapUsed=83790024 external=519775
Memory usage: heapUsed=83792600 external=519775
Memory usage: heapUsed=83795032 external=519775
```
</details>

- Peak heap memory: **85.3 MB**
- Average heap memory: 82.4 MB
- Peak external memory: 31.5 MB
- Average external memory: 15.1 MB


## Testing Plan
- Automated tests for correctness
- Manual testing of performance (code below)

<details>
<summary>Click here to view benchmark code</summary>
```js
// @ts-check

const fs = require('fs');
const logging = require('..');
const fixtures = require('@votingworks/fixtures');
const { join } = require('path');
const { pipeline } = require('stream/promises');

const logger = new logging.Logger(logging.LogSource.VxAdminFrontend);

console.log('Exporting CDF log (pid %d)', process.pid);

function logMemoryUsage() {
  const { heapUsed, external } = process.memoryUsage();
  console.log('Memory usage: heapUsed=%d external=%d', heapUsed, external);
}

setInterval(logMemoryUsage, 1000);

console.time('cdf-export');
pipeline(
  logging.buildCdfLog(
    logger,
    fixtures.electionFamousNames2021Fixtures.electionDefinition,
    fs.createReadStream(join(__dirname, '../fixtures/biglog-100k.log')),
    'machine-id',
    'code-version',
    'system'
  ),
  fs.createWriteStream(join(__dirname, '../fixtures/biglog-100k.cdf.log'))
).then(() => {
  console.timeEnd('cdf-export');
  process.exit(0);
});
```
</details>

## Checklist

- [ ] I have added [logging](https://github.com/votingworks/vxsuite/tree/main/libs/logging) where appropriate to any new user actions, system updates such as file reads or storage writes, or errors introduced.

- [ ] I have added a screenshot and/or video to this PR to demo the change
- [ ] I have added the "user_facing_change" label to this PR to automate an announcement in #machine-product-updates
